2022/Q2/artifacts/pos-model/PoS-model-cubic.md

```diff
@@ -0,0 +1,493 @@
+    var delbonds = compute_total_from_deltas(bonds[delegator_address][validator_address].deltas, cur_epoch + unbonding_length)
+    //check if there are enough bonds
+    //this serves to check that there are bonds (in the docs) and that these are greater than the amount we are trying to unbond
+    if (delbonds >= amount) then
+      //Decrement bond deltas and create unbonds
+      var remain = amount
+      var epoch_counter = cur_epoch + unbonding_length + 1
+      while remain > 0 do
+        epoch_counter = max{epoch | bonds[delegator_address][validator_address].deltas[epoch] > 0 && epoch < epoch_counter}
+        var bond = bonds[delegator_address][validator_address].deltas[epoch_counter]
+        if bond > remain then var unbond_amount = remain
+        else var unbond_amount = bond
+        unbonds[delegator_address][validator_address].deltas[(epoch_counter,cur_epoch+unbonding_length)] += unbond_amount
+        remain -= unbond_amount
+      bonds[delegator_address][validator_address].deltas[cur_epoch+unbonding_length] -= amount
+      //compute new total_deltas and write it at n+unbonding_length
+      var total = read_epoched_field(validators[validator_address].total_deltas, cur_epoch+unbonding_length, 0)
+      validators[validator_address].total_deltas[cur_epoch+unbonding_length] = total - amount
+      //update validator's voting_power, total_voting_power and validator_sets at n+unbonding_length
+      power_before = read_epoched_field(validators[validator_address].voting_power, cur_epoch+unbonding_length, 0)
+      power_after = update_voting_power(validator_address, cur_epoch+unbonding_length)
+      update_total_voting_power(cur_epoch+unbonding_length)
+      update_validator_sets(validator_address, cur_epoch+unbonding_length, power_before, power_after)
+}
+```
+
+```go
+//This function is called by transactions tx_withdraw_unbonds_validator and tx_withdraw_unbonds_delegator
+func withdraw(validator_address, delegator_address)
+{
+  var delunbonds = {<start,end,amount> | amount = unbonds[delegator_address][validator_address].deltas[(start, end)] > 0 && end <= cur_epoch }
+  //substract any pending slash before withdrawing
+  forall (<start,end,amount> in selfunbonds) do
+    var amount_after_slashing = amount
+    forall (slash in slashes[validator_address] s.t. start <= slash.epoch && slash.epoch <= end)
+      amount_after_slashing += amount*slash.rate
+    balance[delegator_address] += amount_after_slashing
+    balance[pos] -= amount_after_slashing
+    //remove unbond
+    unbonds[delegator_address][validator_address].deltas[(start,end)] = 0
+}
+```
+
+```go
+func new_evidence(evidence)
+{
+  //create slash
+  var vpower = read_epoched_field(validators[evidence.validator].total_deltas, evidence.epoch, 0)
+  var slash = Slash{epoch: evidence.epoch, rate: 0, voting_power: vpower}
+  //enqueue slash (Step 1.1 of cubic slashing)
+  append(enqueued_slashes[evidence.epoch + unbonding_epoch], slash)
```

Comment:
I can't find the definition of `enqueued_slashes`

Reply:
Also, what would happen if `evidence.epoch` is too far in the past? (is it possible?)

Reply:
Fixed.